[interim] Ament Lint CLI Utilities (#4908)
ros2/ros2_documentation: source/Tutorials/Advanced/Ament-Lint-For-Clean-Code.rst

Ament Lint CLI Utilities
========================

> **Goal:** Learn how to use ``ament_lint`` and related tools to identify and fix code quality issues

Reviewer: ```suggestion
**Goal:** Learn how to use ``ament_lint`` and related tools to identify and fix code quality issues.
```

**Tutorial level:** Advanced

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

The ``ament`` family of CLI tools are used for CMake-based software development with ROS 2.
Ament ships with a collection of CLI programs that can help users write code that meet the ROS 2 coding standards.

> Using these tools can greatly increase development velocity and help users write ROS applications and core code that meet the doc:`the ROS project's coding standards. <../../The-ROS2-Project/Contributing/Code-Style-Language-Versions>`.

Reviewer: ```suggestion
Using these tools can greatly increase development velocity and help users write ROS applications and core code that meet the doc:`the ROS project's coding standards <../../The-ROS2-Project/Contributing/Code-Style-Language-Versions>`.
```

We recommend that ROS developers familiarize themselves with these tools and use them before submitting their final pull requests. 

Prerequisites
-------------

You should have the ``ament`` packages installed as part of your regular ROS 2 setup.

If you need to install ROS 2, see the :doc:`Installation instructions <../../Installation>`.

Ament Lint CLI Tools
--------------------

1 ``ament_copyright``
^^^^^^^^^^^^^^^^^^^^^

The ``ament_copyright`` CLI can be used to check and update the copyright declaration in ROS source code.
This tool can also be used to check for the presence of an appropriate software license, copyright year, and copyright holders in your source code.

> The ``ament_copyright`` tool works relative to the directory in which it is called, and walks the subdirectories and checks each source file within the directory [And dependencies?].

Reviewer: if i am not mistaken, i do not think it does any work for dependent packages. it walks through the directory recursively and check files.

Author: I was trying to figure this out myself. I tried running the tool in a couple of packages I had sitting around and I got some unexpected results. I think there might be a few unimplemented or poorly worded options in this tool. For example, I couldn't get the `--add-missing` option to work, and the `--list-licenses` flag provides the licenses you can assign, not the licenses in the directory, which is what I understood from reading the help text. This last one is what made me think it was walking the full dependency tree. I was going to take it up with Chris and figure out a plan. 

Anyway, this is an interim PR. I'll fix it up a bit and issue a real PR shortly.

Reviewer 2: All of the `ament_*` tools could likely use some love, as many of the options are rarely invoked.  That said, `ament_copyright` is a particularly odd tool, in that it does different things depending on how it is invoked (with `--list-licenses` or not; that one is my fault), and depending on whether there is one or >1 packages in the repository.  It could probably use a heavy revamp.

The program takes a single optional argument which is a list of directories that should be scanned for the report.
For example, if you wish to scan just source and header files for copyright notices you can call `` ament_copyright ./src ./include``.

1.2 ``ament_copyright`` Options
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``ament_copyright`` supports the following options:

* ``-h, --help`` - shows a help message and exit
* ``--exclude [filename ...]`` - The filenames to exclude.
* ``--add-missing COPYRIGHT_NAME LICENSE`` - Add missing copyright notice and license information using the passed copyright holder and license. ``LICENSE`` passed to this option is the name of the license to be used. A full list of available licenses can be found by calling ``ament_copyright --list-licenses``
* ``--add-copyright-year`` - Add the current year to existing copyright notices.
* ``--list-copyright-names`` - List names of known copyright holders.
* ``--list-licenses`` - List names of known licenses )
* ``--verbose`` - Show all files instead of only the ones with errors / modifications (default: False)

> * ``--xunit-file XUNIT_FILE`` - Generate a xunit compliant XML file (default: None)

Reviewer: i would not explain the options in this tutorial for a couple of reasons.

- maintenance cost: once option is changed or modified, we need to follow the implementation.
- options are explained in the command help, that is the most correct and accurate with user ROS 2 environment.

how about saying that `please see available options with --help options.` instead?

Author: I'm going to push back on this a bit. 

I understand the concerns about maintenance burden and having a single source of truth, but saying that someone needs to install a tool to read their docs is a bit much. As it stands we don't really have documentation related to these tools and we need a place where we can point new users.

I'd be open to just including just a few of the more important options, and collapsing the shared options. 

Would that work?

Reviewer 2: I can see both points of view here.

The real problem with maintenance is that nobody keeps up with this, *and* it differs between ROS 2 distributions (usually we add at least some new options between distributions).  So this will quickly get out-of-date, and will probably be wrong when we backport this article to e.g. Humble.

On the other hand, having the options documented somewhere is nice.

I'm going to say that I don't have a strong opinion one way or the other, so I'll let you all make the call here.

1.3 ``ament_copyright`` Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To check if your ROS package has an appropriate copyright and license file simply call ``ament_copyright`` with no arguments.
Using the ``--verbose`` option will list all checked files.

.. code-block:: console

  ament_copyright --verbose

If you happen to find an issue with one of your files you can address it by calling the following command.

2 ``ament_cppcheck``
^^^^^^^^^^^^^^^^^^^^

The ``ament_cppcheck`` command line tool can be used to quickly perform static analysis of C++ source code files.
`Static analysis <https://en.wikipedia.org/wiki/Static_program_analysis>`_ is the process of automatically reviewing source code files for patterns that can often cause issues after compilation.
Some versions of cpp_check, the underlying utility used by ``ament_cppcheck``, can be rather slow.

> For this reason ``ament_cpp_check`` may be disabled on some systems.

Reviewer: ```suggestion
For this reason ``ament_cppcheck`` may be disabled on some systems.
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default ``ament_cppcheck`` walks the directory in which it is called, including subdirectories and returns a report that lists all of the potential issues in a source code file.
The program takes a single optional argument which is a list of directories that should be scanned for the report.
For example, if you wish to scan just a recently modified file you can call ``ament_cppcheck ./src/my_cpp_file.cpp``.

2.2 ``ament_cppcheck`` Options
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``ament_cppcheck`` supports the following options:

* ``-h, --help``- Show a help message and exit.
* ``--libraries [LIBRARIES ...]`` - Library configurations to load in addition to the standard libraries of C and C++. Each library is passed to cppcheck as '--library=<library_name>'
* ``--include_dirs [INCLUDE_DIRS ...]`` - Include directories for C/C++ files being checked.Each directory is passed to cppcheck as '-I <include_dir>' (default: None)
* ``--exclude [EXCLUDE ...]`` - Exclude C/C++ files from being checked.
* ``--xunit-file XUNIT_FILE`` - Generate a xunit compliant XML file (default: None)
* ``--cppcheck-version`` - Get the cppcheck version, print it, and then exit.

> 2.3 ``ament_cppcheck`` Example

Reviewer: I think this example and corresponding error message association is really good for user to understand what it can do.

   }

Applying ``ament_cpplint`` to this file will yield the following errors:

.. code-block:: console

  example.cpp:0:  No copyright message found.  You should have a line: "Copyright [year] <Copyright Owner>"  [legal/copyright] [5]
  example.cpp:5:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  example.cpp:7:  Tab found; better to use spaces  [whitespace/tab] [1]
  example.cpp:7:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  example.cpp:7:  Missing spaces around =  [whitespace/operators] [4]

4 ``ament_flake8``
^^^^^^^^^^^^^^^^^^

`Flake8 <https://pypi.org/project/flake8/>`_ is a Python tool for linting and style enforcement.
The ``ament_flake8`` command line tool can be used to quickly perform linting of Python source code files using `Flake8 <https://pypi.org/project/flake8/>`_.
This tool will help you locate and fix minor errors and style problems with your ROS Python programs such as trailing whitespace, overly long lines of code, poorly spaced function arguments, and much more!
Unfortunately, at this time, ``ament_flake8`` is unable to to automatically fix these issues.

> Deveopers who whould like to address linting issues can use `Python's Black utility <https://github.com/psf/black>`_ to fix formatting issues automatically. 

Reviewer: (Just FYI) there is a 3rd party package https://github.com/Timple/ament_black, but i am not even sure it is maintained...

Author: I think we should be pointing people to tools that automatically fix smaller formatting errors instead of just pestering them. Perhaps we should build an official tool. I actually left myself a note that this would be a fun little project for Kilted. I'll take a look.

Reviewer 2: Sigh.  This is a long-standing issue.

First of all, specifically for this article, I would not mention the `black` utility.  It is just one of half a dozen replacements for flake8, and it does not format things exactly like flake8 wants.  As this article is meant to be about the ROS ecosystem as it stands today, I don't think we should recommend it.

The broader issue is whether we should switch to one of these other formatters that can do automatic reformatting.  While I'm not fundamentally opposed to that, it is quite a lot more work than people think:

1.  You have to choose which of the half-dozen of these you want to use.  They are all opinionated, and they all have different opinions.  And all of them choose different opinions than flake8 does.
2.  You have to make sure that formatter is available as a binary package on all of our Tier-1 and Tier-2 platforms.  This is trickier than you might think, since a number of them are built in Rust now.  So this step usually starts with "package these 300 Rust crates into Conda, Fedora, EPEL, Debian, and Ubuntu".  This is a prerequisite for any of the steps below.
3.  You need to write the ament tool to run it and interpret the output.
4.  You have to reformat all of the python packages in https://github.com/ros2/ros2/blob/rolling/ros2.repos, at the very least.  That will probably lead to dozens of PRs.
5.  You have to convince the community that this is in their best interest, because this will likely affect quite a number of downstream packages that just choose the default set of linters.

In my opinion, none of this is worth the effort.  This is a lot of work to go from one set of opinions to another set of opinions, neither of which is fundamentally better.  But if somebody wants to go through all of that work, more power to them.

For example, if you wish to scan just source and header files for copyright notices you can call: ``ament_copyright ./src ./include``.

1.2 ``ament_copyright`` Options
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``ament_copyright`` supports the following options:

* ``--add-missing COPYRIGHT_NAME LICENSE`` - Add missing copyright notice and license information using the passed copyright holder and license. ``LICENSE`` passed to this option is the name of the license to be used. A full list of available licenses can be found by calling ``ament_copyright --list-licenses``
* ``--add-copyright-year`` - Add the current year to existing copyright notices.
* ``--list-copyright-names`` - List names of known copyright holders.
* ``--list-licenses`` - List names of known licenses.
* ``--verbose`` - Show all files instead of only the ones with errors / modifications.

1.3 ``ament_copyright`` Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To check if your ROS package has an appropriate copyright and license file simply call ``ament_copyright`` with no arguments.
Using the ``--verbose`` option will list all checked files.

.. code-block:: console

  ament_copyright --verbose

> If you happen to find an issue with one of your files you can address it by calling the following command.

Reviewer 2: I think the rest of this example is missing.

Author: I fixed it up. I am going to run through ament_copyright and submit an issue because I think there might be a few  bugs and / or poorly specified options in there. That's why the example trailed off.

Reviewer 2: Yeah, I'm sure there are.  A new issue about that is great.

2 ``ament_cppcheck``
^^^^^^^^^^^^^^^^^^^^

> The ``ament_cppcheck`` command line tool can be used to quickly perform static analysis of C++ source code files.

Reviewer 2: ```suggestion
The ``ament_cppcheck`` command line tool can be used to perform static analysis of C++ source code files.
```

`Static analysis <https://en.wikipedia.org/wiki/Static_program_analysis>`_ is the process of automatically reviewing source code files for patterns that can often cause issues after compilation.

> Some versions of cpp_check, the underlying utility used by ``ament_cppcheck``, can be rather slow.

Reviewer 2: ```suggestion
Some versions of `cppcheck <https://github.com/danmar/cppcheck>`__, the underlying utility used by ``ament_cppcheck``, can be rather slow.
```

Ament Lint CLI Utilities
========================

**Goal:** Learn how to use ``ament_lint`` and related tools to identify and fix code quality issues.

**Tutorial level:** Advanced

**Time:** 10 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

> The ``ament`` family of CLI tools are used for CMake-based software development with ROS 2.

Reviewer 2: The way that this is structured is that the `ament` tools can be used from *any* build system, since they are just python scripts.  So, for instance, it is entirely possible to run the ament tools from CMake, from pytest, from Meson, etc.  We *also* provide a set of `ament_cmake` helpers to make using them specifically from CMake easier.  But I do think it is worth pointing out that distinction.

Author: Added a sentence to clarify.

Ament ships with a collection of CLI programs that can help users write code that meet the ROS 2 coding standards.
Using these tools can greatly increase development velocity and help users write ROS applications and core code that meet the doc:`the ROS project's coding standards <../../The-ROS2-Project/Contributing/Code-Style-Language-Versions>`.

> We recommend that ROS developers familiarize themselves with these tools and use them before submitting their final pull requests.

Reviewer 2: ```suggestion
We recommend that ROS developers familiarize themselves with these tools and use them before submitting their pull requests.
```

Prerequisites
-------------

You should have the ``ament`` packages installed as part of your regular ROS 2 setup.

If you need to install ROS 2, see the :doc:`Installation instructions <../../Installation>`.

Ament Lint CLI Tools
--------------------

All ament linting tools use a similar CLI pattern.

> They take in a directory, a list of directories, file, or list of files, analyze the input files, and generate report.

Reviewer 2: ```suggestion
They take in a directory, a list of directories, file, or list of files, analyze the input files, and generate a report.
```

All ament linting tools have the following built-in options.
**The most up to date and accurate documentation for a given ament tool can be found by using the tools built in ``--help`` functionality.**

* ``-h, --help`` - shows a help message and exit.
  The built-in help messages usually have the most accurate and up-to-date documentation of the tool.
* ``--exclude [filename ...]`` - The filenames to exclude from analysis, including wildcards.
* ``--xunit-file XUNIT_FILE`` - Generate a `xunit <https://xunit.net/>`_ compliant XML file.

>   These files are most commonly used by IDEs to automate the ingestion of external development tools.

Reviewer 2: ```suggestion
  These files are most commonly used by IDEs and CI to automate the ingestion of test results.
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Let's create a simple C++ program named ``example.cpp``.
We will add a few lines of code that violate coding standards

.. code-block:: cpp

  int main()
  {
       int a = 10;
       int b = 10;
       int c = 0;  
       if( a == b)  {
 	  c=a;}  
       return 0;
   }

Applying ``ament_cpplint`` to this file will yield the following errors:

.. code-block:: console

  example.cpp:0:  No copyright message found.  You should have a line: "Copyright [year] <Copyright Owner>"  [legal/copyright] [5]
  example.cpp:5:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  example.cpp:7:  Tab found; better to use spaces  [whitespace/tab] [1]
  example.cpp:7:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  example.cpp:7:  Missing spaces around =  [whitespace/operators] [4]

4 ``ament_flake8``
^^^^^^^^^^^^^^^^^^

`Flake8 <https://pypi.org/project/flake8/>`_ is a Python tool for linting and style enforcement.
The ``ament_flake8`` command line tool can be used to quickly perform linting of Python source code files using `Flake8 <https://pypi.org/project/flake8/>`_.

> This tool will help you locate and fix minor errors and style problems with your ROS Python programs such as trailing whitespace, overly long lines of code, poorly spaced function arguments, and much more!
> Unfortunately, at this time, ``ament_flake8`` is unable to to automatically fix these issues.

Reviewer 2: This isn't an unfortunate oversight; the upstream developers of flake8 have said that they never intend to implement automatic code reformatting in flake8 itself.  So we should reword this to make it clear that this one only detects errors and cannot fix them:

```suggestion
This tool will help you locate minor errors and style problems with your ROS Python programs such as trailing whitespace, overly long lines of code, poorly spaced function arguments, and much more!
Note, however, that ``flake8`` and ``ament_flake8`` cannot automatically reformat code to fix these issues.
```

> Deveopers who whould like to address linting issues can use `Python's Black utility <https://github.com/psf/black>`_ to fix formatting issues automatically.

Reviewer 2: ```suggestion
```

      extra_long =(thisIsAVariableNameThatIsWayTooLongLongLong*thisIsAVariableNameThatIsWayTooLongLongLong )
      ^

  example.py:5:17: E225 missing whitespace around operator
      extra_long =(thisIsAVariableNameThatIsWayTooLongLongLong*thisIsAVariableNameThatIsWayTooLongLongLong )
                  ^

  example.py:5:100: E501 line too long (106 > 99 characters)
      extra_long =(thisIsAVariableNameThatIsWayTooLongLongLong*thisIsAVariableNameThatIsWayTooLongLongLong )
                                                                                                     ^

  example.py:5:105: E202 whitespace before ')'
      extra_long =(thisIsAVariableNameThatIsWayTooLongLongLong*thisIsAVariableNameThatIsWayTooLongLongLong )
                                                                                                          ^

  1     E202 whitespace before ')'
  1     E225 missing whitespace around operator
  1     E231 missing whitespace after ','
  1     E501 line too long (106 > 99 characters)
  1     F841 local variable 'extra_long' is assigned to but never used

  1 files checked
  5 errors

  'E'-type errors: 4
  'F'-type errors: 1

  Checked files:

  * example.py

> If you have installed Python's `Black utility <https://github.com/psf/black>`_ it is possible to address these issues directly by calling ``black example.py.``
> It is also worth noting that there is a community contributed ament tool for Python Black called ``ament_black`` that can be found on `ROS Index <https://index.ros.org/p/ament_black/>`_. 

Reviewer 2: ```suggestion
```

  Checked files:

  * example.py

If you have installed Python's `Black utility <https://github.com/psf/black>`_ it is possible to address these issues directly by calling ``black example.py.``

5 ``ament_uncrustify``
^^^^^^^^^^^^^^^^^^^^^^

`Uncrustify <https://github.com/uncrustify/uncrustify>`_ is a C++ linting tool, similar to ``ament_cpplint``, that has the advantage that it can **automatically fix** the issues it finds!

> This tool will help you locate and fix minor errors and style problems with your C++ ROS  programs such as trailing whitespace, overly long lines of code, poorly spaced function arguments, and much more!

Reviewer 3: ```suggestion
This tool will help you locate and fix minor errors and style problems with your C++ ROS programs such as trailing whitespace, overly long lines of code, poorly spaced function arguments, and much more!
```
